### PR TITLE
Introduce tutorials

### LINKED ISSUE #41: Make tutorials editiable via client

### PR DESCRIPTION
**Pull request by axel1200**
Fixes #41

### CODE REVIEW

File: src/main/java/com/faforever/moderatorclient/ui/TutorialAddController.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        close();
+        if(onSaveRunnable!=null){
+            onSaveRunnable.run();
+        }
+    }
+
+    private boolean validate() {
+        if(titleField.getText().isEmpty()){
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/TutorialAddController.java]: Please catch all errors in a List and concat all of them in a stream to show all of them at once. There are examples in other Controllers.

File: src/main/java/com/faforever/moderatorclient/ui/TutorialAddController.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            try{
+                Integer.parseInt(mapField.getText());
+            }catch (Exception e){
+                return error("Map id must be a valid Number");
+            }
+        }
+        return true;
+    }
+
+    private boolean error(String message) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/TutorialAddController.java]: Use ViewHelper.errorDialog instead

File: src/main/java/com/faforever/moderatorclient/ui/MessagesController.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import javafx.stage.Stage;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import javax.inject.Inject;
+import java.util.function.Predicate;
+
+@Component
+@Slf4j
+public class MessagesController {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/MessagesController.java]: implements Controller\<HBox\>

File: src/main/java/com/faforever/moderatorclient/ui/MessagesController.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    @Inject
+    public MessagesController(MessagesService messagesService, UiService uiService) {
+        this.messagesService = messagesService;
+        this.uiService = uiService;
+    }
+
+    @FXML
+    public void initialize() {
+        ViewHelper.buildTutorialTable(messageTableView, messagesService, log);
+        onRefreshMessages();
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/MessagesController.java]: Don't invoke loading of data in @FXML initialize.  Call them manually from MainController. This is due to error handling. We want to throw errors first after the user passed the login window (this is new behavior in develop).

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java [hunk 5]
```diff
@@ -750,6 +756,236 @@ public static void buildNotesTableView(TableView<UserNoteFX> tableView, Observab
         tableView.getColumns().add(updateTimeColumn);
     }
 
+    public static void buildTutorialTable(TableView<TutorialFx> tutorialTableView, TutorialService tutorialService, Logger log, Runnable refresh) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: Please add copy & paste behavior as is done now in all controls in develop. See examples there.

File: src/main/java/com/faforever/moderatorclient/ui/main_window/TutorialController.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        filterTutorials = new FilteredList(tutorialList);
+    }
+
+    @FXML
+    public void initialize(){
+        Runnable refreshTutorials = this::onRefreshTutorials;
+        Runnable refreshCategorys = this::onRefreshCategorys;
+        ViewHelper.buildTutorialTable(tutorialTableView, tutorialService, log, refreshTutorials);
+        ViewHelper.buildCategoryTable(categoryTableView, tutorialService, refreshCategorys);
+        onRefreshTutorials();
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/TutorialController.java]: Same as above. Don't load data in @FXML initializers. Call them manually from MainController.

File: src/main/resources/application.yml [hunk 7]
```diff
@@ -1,8 +1,8 @@
 faforever:
   api:
-    base-url: https://api.faforever.com
-    client-id: 05209670-36c5-40ad-b581-e71b7b49bd1a
-    client-secret: 370965a5-e7f2-4fe1-9225-e4517476b6a5
+    base-url: http://localhost:8010
```
Comment (Brutus5000) [inline src/main/resources/application.yml]: Setup your run configurations. You can use the dev-profile. Revert this.

File: src/main/resources/ui/mainWindow.fxml [hunk 8]
```diff
@@ -13,6 +13,12 @@
         <Tab fx:id="avatarsTab" closable="false" text="Avatars"/>
         <Tab fx:id="recentActivityTab" closable="false" text="Recent activity"/>
         <Tab fx:id="domainBlacklistTab" closable="false" text="Domain Blacklist"/>
+        <Tab closable="false" text="Tutorials">
+            <fx:include source="main_window/tutorial.fxml"/>
```
Comment (Brutus5000) [inline src/main/resources/ui/mainWindow.fxml]: Don't include. Use fx:id in Tab and load it as all other tabs are loaded.

File: src/main/resources/ui/mainWindow.fxml [hunk 9]
```diff
@@ -13,6 +13,12 @@
...
+        </Tab>
+        <Tab closable="false" text="Messages">
+            <fx:include source="main_window/messages.fxml"/>
```
Comment (Brutus5000) [inline src/main/resources/ui/mainWindow.fxml]: Same here.

File: src/main/resources/ui/mainWindow.fxml [hunk 10]
```diff
@@ -13,6 +13,12 @@
...
+        </Tab>
         <Tab closable="false" text="Bans">
             <fx:include source="main_window/bans.fxml"/>
```
Comment (Brutus5000) [inline src/main/resources/ui/mainWindow.fxml]: And here.

**Comment by axel1200** [comment]
Well the api does not work 100 % so i will need to repare the java api first

**Comment by axel1200** [comment]
Note that on production adding messages may be problematic as newest api is not yet deployed

**Comment by axel1200** [comment]
@Brutus5000 could you review this so that I can make the ladder voting stuff on top of this(or rebase it on top of this)

**Comment by axel1200** [comment]
Was there not something in the java api which caused the 423 Locked???

**Comment by axel1200** [comment]
If you enter a String that is too long in language or Region is does not work there is a limit to it. What do I do make a pop up?? Anyway moderators should enter valid data. So not sure if that is too much of a problem

**Comment by Brutus5000** [comment]
Here is an example for validation: https://github.com/FAForever/faf-moderator-client/blob/develop/src/main/java/com/faforever/moderatorclient/ui/BanInfoController.java#L146

**Comment by axel1200** [comment]
I did validation and I also have a pop up if user enters more than two characters into table cell. Just try your self I would say

**Comment by Brutus5000** [comment]
Why did you add `dialogOnError` to the `FafApiFailGetEvent`? When would you not throw this as an error popup?

**Comment by axel1200** [comment]
@Brutus5000 When I expect one 😬 or if I wanna have my own messgage. So I wanna handle the Exception my self. For example if you enter a wrong map version id: I dont wanna have the user read a page of text telling him something went wrong etc. and at the end the is completely confused. I wanna just tell him this map version id he entered is wrong.

**Comment by axel1200** [comment]
Well the exception was my check XD, but yeah I can do it via filter=id=={id}

**Comment by Brutus5000** [comment]
Ok so I'm finally able to create a tutorial record.

What i noticed now:
- [ ] When I update the map version id to a not-existing one, instead of an error message I am just stuck in the table cell again (like we had before with the region probably because of underneath exception)
- [ ] All  keys don't have a database constraint to messages. As this is a little of the chicken-egg problem I will not enforce this by now. Would like to hear@micheljung opinion.

**Comment by axel1200** [comment]
I followed previous examples concerning messages. Maybe we should make an db issue on it. But I can add one thing to that. So the only thing that the constraint would change is that you can not enter none existing message keys but otherwise it would not speed up querries etc.. The java-api loades all messages into memory and refreshes them if new ones are created.

**Comment by axel1200** [comment]
Concerning the missing error you are right, I will quickly add that.

**Pull request merged**